lib/execution_engine2/sdk/EE2Runjob.py

```diff
@@ -457,6 +499,166 @@ def _check_job_arguments(self, jobs, has_parent_job=False):
+        self.logger.error(msg, exc_info=True, stack_info=True)
+        raise RetryFailureException(msg)
+
+    def _validate_retry_presubmit(self, job_id, as_admin):
+        """Validate retry request before attempting to contact scheduler"""
+        # Check to see if you still have permissions to the job and then optionally the parent job id
+        job = self.sdkmr.get_job_with_permission(
+            job_id, JobPermissions.WRITE, as_admin=as_admin
+        )  # type: Job
+        job_input = job.job_input  # type: JobInput
+
+        parent_job = None
+        if job_input.parent_job_id:
+            parent_job = self.sdkmr.get_job_with_permission(
+                job_input.parent_job_id, JobPermissions.WRITE, as_admin=as_admin
+            )
+
+        if job.batch_job:
+            raise CannotRetryJob(
+                "Cannot retry batch job parents. Must retry individual jobs"
+            )
+
+        if not self._retryable(job.status):
+            raise CannotRetryJob(
+                f"Can only retry a cancelled or errored job_id:{job_id} status:{job.status}"
+            )
+
+        return job, job_input, parent_job
+
+    def retry(self, job_id: str, as_admin=False) -> Dict[str, Optional[str]]:
+        """
+        #TODO Add new job requirements/cgroups as an optional param
+        :param job_id: The main job to retry
+        :param as_admin: Run with admin permission
+        :return: The child job id that has been retried
+        """
+
+        job, job_input, parent_job = self._validate_retry_presubmit(
+            job_id=job_id, as_admin=as_admin
+        )
+
+        # Cannot retry a retried job, you must retry the retry_parent
+        if job.retry_parent:
+            return self.retry(str(job.retry_parent), as_admin=as_admin)
+
+        # Get run job params from db, and inject parent job id, then run it
+        run_job_params = self._get_run_job_params_from_existing_job(
+            job, user_id=self.sdkmr.user_id
+        )
+        # Submit job to job scheduler or fail and not count it as a retry attempt
+        run_job_params[_PARENT_RETRY_JOB_ID] = job_id
+        retry_job_id = self.run(params=run_job_params, as_admin=as_admin)
+
+        # Save that the job has been retried, and increment the count. Notify the parent(s)
+        # 1) Notify the retry_parent that it has been retried
+        try:
+            job.modify(inc__retry_count=1, set__retried=True)
+        except Exception as e:
+            self._db_update_failure(
+                job_that_failed_operation=str(job.id),
+                job_to_abort=retry_job_id,
+                exception=e,
+            )
+
+        # 2) Notify the parent container that it has a new child..
+        if parent_job:
+            try:
+                parent_job.modify(push__child_jobs=retry_job_id)
+            except Exception as e:
+                self._db_update_failure(
```

Comment:
if this one fails, then we abort the retry job, but the parent is still marked as retried and has the retry count bumped, and the retry job still has the retry parent job field filled in. Is that an issue?

Reply:
I don't think it's an issue

Reply:
Hmm, why are you convinced of that? If I'm a FE coder and I find a job that says it's been retried, but the retried job isn't in the set of jobs for the batch, isn't that an issue?

Reply:
Seems like the frontend would already know about it. If it didn't know about it, it wouldn't display it. If it knew about the job, it would display it. If it says it has been retried, but the retry doesn't exist, then you can assume it failed.

Reply:
@ialarmedalien can you give your take on this given you're going to have to deal with it? I'm worried I may be being too picky

Reply:
Because of potential issues with synchronicity, missed updates, etc., the narrative will always assume that that ee2 is the ultimate source of truth, so anywhere that there are discrepancies in the ee2 data is going to potentially cause problems for the narrative. The plan is that after requesting a retry, the narrative BE will request an update on the parent job and from that, get the list of child jobs and update those, too.

Ideally there would be some sort of db integrity check so that it's impossible to get into a state where you have batch job containers that have child jobs that don't exist / jobs with retry parents where the reciprocal info has not been added / etc., but I'd guess that is not the case -- correct?

Reply:
IMO it'd be better to put all of the inter-job link making into a single transaction, so that either you get all the job links made (`retry_parent`, child jobs updated, retry count updated) or you get none. Having an intermediate state means you're making any clients do a load of error checking and inference about job structures, and the db can't really be trusted.

Reply:
Unfortunately Mongo 3.6 doesn't support transactions. It's possible to invent a transaction-like mechanism (https://github.com/kbase/sample_service/blob/master/lib/SampleService/core/storage/arango_sample_storage.py#L8-L27), but from experience I can say it's a substantial chunk of work

Reply:
Hmmm, that is a nuisance. Looks like the options are

- implement db integrity checks and [two-phase commits](https://docs.mongodb.com/v3.6/tutorial/perform-two-phase-commits/) for making the relationships between a job, its retry parent, and the batch container
- accept that the db info may be incomplete and write workarounds into the clients
- (upgrade to Mongo 4.4 for proper transaction support)

If clients have to work around potentially incomplete db data, giving jobs a `batch_id` (or `parent_id` or something similar) field would make it easier to see if a job is part of a batch. For the retry data, I'd say it's slightly preferable to add the child job to the batch container's child jobs list _before_ amending the retry count and retried fields on the retry parent job.

Reply:
We agreed to table this for now and discuss in an ADR: https://github.com/kbase/execution_engine2/pull/367/files#diff-370ecf4c171a1ae480f063fbe22e0dbd2c9a418ba7f8cf51dce663887c74196fR76

Reply:
Nice ideas.  Added to the ADR.
The jobs  do indeed have a parent_id which makes it possible to see if the job is part of a batch. 
Added the job to child jobs list before amending the retry count. The retried field is now deleted as per https://github.com/kbase/execution_engine2/pull/383#discussion_r636534319